src/nbl/asset/utils/CSPIRVIntrospector.cpp

```diff
@@ -13,472 +13,528 @@ namespace nbl::asset
-    const std::span<const ICPUShader::SSpecInfo> infos)
-{
-#if 0 // rewrite for sanity
-    core::vector<core::vector<SPushConstantRange>> ranges(infos.size());
-    {
-        auto introspection = introspections;
-        auto r = ranges.begin();
-        for (auto info : infos)
+                    return val * specConstantFound->second;
+                }
+
+                return val * arrInfo.value;
+            };
+
+        const auto& introBindingInfos = stageData->getDescriptorSetInfo(i);
+        for (const auto& stageIntroBindingInfo : introBindingInfos)
         {
-            auto& pc = (*introspection)->pushConstant;
-            if (pc.present)
+            CPipelineIntrospectionData::SDescriptorInfo descInfo;
+            descInfo.binding = stageIntroBindingInfo.binding;
+            descInfo.type = stageIntroBindingInfo.type;
+            descInfo.stageMask = stageData->m_shaderStage;
+            if (stageIntroBindingInfo.isArray())
             {
-                const auto shaderStage = info.shader->getStage();
-
-                r->reserve(100u);
-                auto& members = pc.info.members;
-                r->push_back({shaderStage, members.array[0].offset, members.array[0].size});
-                for (uint32_t i = 1u; i < members.count; ++i)
+                if (stageIntroBindingInfo.isRunTimeSized())
+                {
+                    descInfo.count = 0u;
+                    descInfo.stride = std::accumulate(stageIntroBindingInfo.count.begin(), stageIntroBindingInfo.count.end(), 1u, multiplyArrayDimensions);
+                    if (descInfo.stride == 0u)
+                        return false;
+                }
+                else
                 {
-                    auto& last = r->back();
-                    if (members.array[i].offset == (last.offset + last.size))
-                        last.size += members.array[i].size;
-                    else
-                        r->push_back({shaderStage, members.array[i].offset, members.array[i].size});
+                    descInfo.count = std::accumulate(stageIntroBindingInfo.count.begin(), stageIntroBindingInfo.count.end(), 1u, multiplyArrayDimensions);
+                    descInfo.stride = descInfo.count;
+                    if (descInfo.stride == 0u)
+                        return false;
                 }
             }
-            introspection++;
-            r++;
+            else
+            {
+                descInfo.count = 1u;
+                descInfo.stride = 0u;
+            }
+
+            const auto& pplnIntroDataFoundBinding = m_descriptorSetBindings[i].find(descInfo);
+            if (pplnIntroDataFoundBinding != m_descriptorSetBindings[i].end())
+            {
+                if (pplnIntroDataFoundBinding->type != stageIntroBindingInfo.type)
+                    return false;
+                if (pplnIntroDataFoundBinding->count != descInfo.count)
+                    return false;
```

Comment:
having non-matching counts is okay, you just `max` their counts

Whats important is to be in agreement whether the binding is runtime sized or not 
Unless there's no iff relationship, then you need to handle "upgrading" a regular binding to a runtime sized one, while also max-ing the counts (because the introspection returned count is now a lower-bound on the size of the runtime array).

Reply:
Also you're not checking whether all the Stage's bindings agree with itself (because there can be resource aliasing) , this why you either need to:
- copy initialize `descriptorsToMerge = m_descriptorSetBindings` and do all checks against that (then you'll check further stage bindings against previous stage bindings), or
- turn this validation from `find` to `debug <<` into a lambda and do two lookups, one in `decriptorsToMerge` and one in `m_descriptorSetBindings`